File: #74_Search_a_2D Matrix/Solution.java
public class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
    	if (matrix == null) return false;
    	int rows = matrix.length;
    	int cols = matrix[0].length;
    	int min = matrix[0][0];
    	int max = matrix[rows - 1][cols - 1];
    	if (target < min || target > max) return false;
    	/* binary search for rows which target belongs to */
    	int lo = 0; int hi = rows - 1; int mid = (lo + hi) / 2;
    	int row = -1;
    	while (true) {
    		if (matrix[mid][0] <= target && matrix[mid][cols - 1] >= target) {
    			row = mid; break;
    		}
    		if (matrix[mid][0] > target) {
    			hi = mid - 1;
    			if (lo > hi || lo >= rows || hi < 0) return false;
    			mid = (lo + hi) / 2;
    		}
    		if (matrix[mid][cols - 1] < target) {
    			lo = mid + 1;
    			if (lo > hi || lo >= rows || hi < 0) return false;
    			mid = (lo + hi) / 2;
    		}
    	}
    	/* binary search for specific row */
    	lo = 0; hi = cols - 1; mid = (lo + hi) / 2;
    	while(true) {
    		if (matrix[row][mid] == target) return true;
    		if (matrix[row][mid] < target) {
    			lo = mid + 1;
    			if (hi < 0 || lo > hi || lo >= cols) return false;
    			mid = (lo + hi) / 2;
    		}
    		if (matrix[row][mid] > target) {
    			hi = mid - 1;
    			if (hi < 0 || lo > hi || lo >= cols) return false;
    			mid = (lo + hi) / 2;
    		}
    	}
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[][] test = {
    		{1},
    		{3}
    	};
    	System.out.println(sol.searchMatrix(test, 3));
    }
}

File: #74_Search_a_2D Matrix/Solution.py
'''
Perform binary search on both columns and rows
time: O(log(rows) + log(cols))
'''

class Solution:
    # @param {integer[][]} matrix
    # @param {integer} target
    # @return {boolean}
    def searchMatrix(self, matrix, target):
        if not matrix or not matrix[0]:
        	return False
        row = self.binarySearch([matrix[x][0] for x in range(len(matrix))], target)
        if row < 0:
        	return False
        col = self.binarySearch(matrix[row], target)
        if col < 0:
        	return False
        return matrix[row][col] == target

    def binarySearch(self, nums, target):
    	lo, hi = 0, len(nums) - 1
    	while lo <= hi:
    		mid = lo + (hi - lo) / 2
    		if nums[mid] == target:
    			return mid
    		if nums[mid] < target:
    			lo = mid + 1
    		else:
    			hi = mid - 1
    	return hi

sol = Solution()
t = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
print sol.searchMatrix(t, 10)